src/rt/array/capacity.d

```diff
@@ -0,0 +1,197 @@
+ * that the slice can grow to before the underlying array must be
+ * reallocated or extended.
+ *
+ * If an append must reallocate a slice with no possibility of extension, then
+ * `0` is returned. This happens when the slice references a static array, or
+ * if another slice references elements past the end of the current slice.
+ *
+ * Note: The _capacity of a slice may be impacted by operations on other slices.
+ */
+@property size_t capacity(T)(T[] arr) pure nothrow @trusted
+{
+    return _d_arraysetcapacity(typeid(T[]), 0, cast(void[]*)&arr);
+}
+
+///
+@safe unittest
+{
+    //Static array slice: no capacity
+    int[4] sarray = [1, 2, 3, 4];
+    int[]  slice  = sarray[];
+    assert(sarray.capacity == 0);
+    //Appending to slice will reallocate to a new array
+    slice ~= 5;
+    assert(slice.capacity >= 5);
+
+    //Dynamic array slices
+    int[] a = [1, 2, 3, 4];
+    int[] b = a[1 .. $];
+    int[] c = a[1 .. $ - 1];
+    debug(SENTINEL) {} else // non-zero capacity very much depends on the array and GC implementation
+    {
+        assert(a.capacity != 0);
+        assert(a.capacity == b.capacity + 1); //both a and b share the same tail
+    }
+    assert(c.capacity == 0);              //an append to c must relocate c.
+}
+
+/**
+ * Reserves capacity for a slice. The capacity is the size
+ * that the slice can grow to before the underlying array must be
+ * reallocated or extended.
+ *
+ * Returns: The new capacity of the array (which may be larger than
+ * the requested capacity).
+ */
+size_t reserve(T)(ref T[] arr, size_t newcapacity) pure nothrow @trusted
+{
+    if (__ctfe)
+        return newcapacity;
+    else
+        return _d_arraysetcapacity(typeid(T[]), newcapacity, cast(void[]*)&arr);
+}
+
+///
+@safe unittest
+{
+    //Static array slice: no capacity. Reserve relocates.
+    int[4] sarray = [1, 2, 3, 4];
+    int[]  slice  = sarray[];
+    auto u = slice.reserve(8);
+    assert(u >= 8);
+    assert(&sarray[0] !is &slice[0]);
+    assert(slice.capacity == u);
+
+    //Dynamic array slices
+    int[] a = [1, 2, 3, 4];
+    a.reserve(8); //prepare a for appending 4 more items
+    auto p = &a[0];
+    u = a.capacity;
+    a ~= [5, 6, 7, 8];
+    assert(p == &a[0]);      //a should not have been reallocated
+    assert(u == a.capacity); //a should not have been extended
+}
+
+// https://issues.dlang.org/show_bug.cgi?id=12330, reserve() at CTFE time
+@safe unittest
+{
+    int[] foo() {
+        int[] result;
+        auto a = result.reserve = 5;
+        assert(a == 5);
+        return result;
+    }
+    enum r = foo();
+}
+
+// Issue 6646: should be possible to use array.reserve from SafeD.
+@safe unittest
+{
+    int[] a;
+    a.reserve(10);
+}
+
+// HACK:  This is a lie.  `_d_arrayshrinkfit` is not `nothrow`, but this lie is necessary
+// for now to prevent breaking code.
+private extern (C) void _d_arrayshrinkfit(const TypeInfo ti, void[] arr) nothrow;
```

Comment:
This important note was absent in `object.d`.  I added it here for full transparency.